src/mapreduce.jl

```diff
@@ -160,25 +160,76 @@ end
 ## reducedim ##
 ###############
 
-@inline reducedim(op, a::StaticArray, ::Val{D}) where {D} = mapreducedim(identity, op, a, Val{D})
-@inline reducedim(op, a::StaticArray, ::Val{D}, v0) where {D} = mapreducedim(identity, op, a, Val{D}, v0)
+@inline reducedim(op, a::StaticArray, ::Type{Val{D}}) where {D} = mapreducedim(identity, op, a, Val{D})
+@inline reducedim(op, a::StaticArray, ::Type{Val{D}}, v0) where {D} = mapreducedim(identity, op, a, Val{D}, v0)
 
 #######################
 ## related functions ##
 #######################
 
 # These are all similar in Base but not @inline'd
-@inline sum(a::StaticArray{<:Any, T}) where {T} = reduce(+, zero(T), a)
-@inline sum(f::Base.Callable, a::StaticArray) = mapreduce(f, +, a)
-@inline prod(a::StaticArray{<:Any, T}) where {T} = reduce(*, one(T), a)
-@inline count(a::StaticArray{<:Any, Bool}) = reduce(+, 0, a)
-@inline all(a::StaticArray{<:Any, Bool}) = reduce(&, true, a)  # non-branching versions
-@inline any(a::StaticArray{<:Any, Bool}) = reduce(|, false, a) # (benchmarking needed)
+#
+# Implementation notes:
+#
+# 1. When providing an initial value v0, note that its location is different in reduce and
+# reducedim: v0 comes earlier than collection in reduce, whereas it is the last argument in
+# reducedim.  The same difference exists between mapreduce and mapreducedim.
+#
+# 2. mapreduce and mapreducedim usually do not take initial value v0, because we don't
+# always know the return type of an arbitrary mapping function f.  (We usually want to use
+# some initial value such as one(T) or zero(T) as v0, where T is the return type of f, but
+# if users provide type-unstable f, its return type cannot be known.)  Therefore, mapped
+# versions of the functions implemented below usually require the collection to have at
+# least two entries.
+#
+# 3. Exceptions are the ones that require Boolean mapping functions.  For example, f in
+# all and any must return Bool, so we know the appropriate v0 is true and false,
+# respectively.  Therefore, all(f, ...) and any(f, ...) are implemented by mapreduce(f, ...)
+# with an initial value v0 = true and false.
+#
+# 4. Some implementations (e.g., count(a, Val{D})) are commented out because corresponding
+# Base functions (e.g., count(a, D)) do not exist yet.
```

Comment:
Yet? Meaning they are mooted for v0.7?

(We are allowed to be a bit more agile and move a bit beyond `Base` wherever it makes sense, so long as we don't break anything that tends to work for `AbstractArray`)